### PR TITLE
add munge signing mechanism

### PR DESCRIPTION
**Pull request by garlick**
This PR implements a `munge` signing mechanism, following the plan outlined in a comment in #59.

I still need to borrow @dun's sharness-fu as discussed in #55, but thought I'd get this PR going anyway.

### CODE REVIEW

**Comment by garlick** [comment]
Looks like travis has munge 0.5.11-1ubuntu1.1, which doesn't have the `--pid-file` option, which how the old version of 03-munged.sh stops the daemon.  
```
munged: Error: Invalid option "--pid-file=/home/travis/build/flux-framework/flux-security/t/trash-directory.t1002-sign-munge/munged.pid.29318"
not ok 1 - start munged
```
I probably just need to build a newer munge in travis...

**Comment by dun** [comment]
@garlick: `op_sign` uses a SHA-1 hash. I recommend using SHA-256 since NIST has [deprecated](https://csrc.nist.gov/projects/hash-functions/nist-policy-on-hash-functions) the use of SHA-1 for most things.

For hash agility, I suggest tagging the hash string encapsulated within the MUNGE credential with the algorithm used to create it.  It'll make it easier to migrate to a different hash in the future should the need arise.

**Comment by garlick** [comment]
Thanks for those suggestions @dun!  Will do.

I think I finally got the sharness tests working correctly with munge-on-the-side.  Hopefully travis agrees!

**Comment by garlick** [comment]
@dun's suggestions implemented, some earlier development squashed, and coverage should be better. 

This will need further squashing before merging.  I'll be offline for a while - train is almost here.

**Comment by garlick** [comment]
Fixed one clang error and squashed.  I think maybe this is ready?

**Comment by grondo** [comment]
Been playing with this a bit on my desktop. Something weird happens when the `[sign.munge]` table is empty:
```
grondo@peep:~/git/flux-security.git/t/src$ echo foo| ./sign
sign: flux_sign_wrap: sign-munge-sign: MUNGE socket name is undefined
grondo@peep:~/git/flux-security.git/t/src$ vim s.toml
grondo@peep:~/git/flux-security.git/t/src$ cat s.toml
[sign]
 max-ttl = 60
 default-type = "munge"
 allowed-types = [ "munge" ]
grondo@peep:~/git/flux-security.git/t/src$ echo foo| ./sign
dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZXJpZABpNjg4NQA=.Zm9vCg==.MUNGE:AwQDAAANQQ0JEnsr9UDH5i41+CCekh5/QbnGh3EmtnrwC2anxe/PAjVDNCYOacA59aoJEOZShIhqZwKE/0riXj5Yy0Xw7pLrvzXpLuRA3rpS8oL5hdxWjTVHuf1ab7R0mc2yoMwlQY3FX7r3zdBpPGAt+llXZkpUqA==:
```

Otherwise seems to work well so far with system installed munge package on my system!

**Comment by garlick** [comment]
Oh I see what that is - let me fix.

**Comment by garlick** [comment]
Just pushed a fix and squashed it.

cf_string(NULL) returns an empty string.  That was by design, but it seems counterintuitive.  I'll open a bug and we can decide if it should be fixed.

**Comment by grondo** [comment]
I'd like @dun's ACK on this, but I think he's out for the next few days. I'll poke at the implementation just a bit more on my desktop, but otherwise, probably OK to merge now once Travis is green.

**Comment by garlick** [comment]
Thanks!

File: .travis.yml [hunk 1]
```diff
@@ -29,7 +29,11 @@ before_install:
   - curl -L -O --insecure https://github.com/jedisct1/libsodium/releases/download/1.0.10/libsodium-1.0.10.tar.gz
   - tar -xf libsodium-1.0.10.tar.gz
   - (cd libsodium-1.0.10 && ./configure --prefix=$HOME/local && make && make install)
+  - curl -L -O --insecure https://github.com/dun/munge/releases/download/munge-0.5.13/munge-0.5.13.tar.xz
```
[the remark this answers is not in the document]
Comment (garlick) [reply .travis.yml]: Thanks for the review!

When testing a proposed PR, a failure is usually interpreted as a problem with the PR.  If (though unlikely) something changed in upstream munge that broke our master and hence derived PR's, it would be a confusing way to find out about it.

Munge has a pretty stable API, and I would be surprised if the author did not meticulously follow shared library versioning rules for any changes, so I'd say the more likely scenario for breakage is that munged options would change, and that would break tests, not the deployed product.

For those reasons, pinning it seems right to me.

File: src/lib/sign_munge.c [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+{
+    struct sign_munge *sm = flux_security_aux_get (ctx, auxname);
+    BYTE digest[SHA256_BLOCK_SIZE + 1] = { HASH_TYPE_SHA256 };
+    SHA256_CTX shx;
+    char *cred;
+    munge_err_t e;
+
+    assert (sm != NULL);
+    sha256_init (&shx);
+    sha256_update (&shx, (const BYTE *)input, strlen (input));
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/lib/sign_munge.c]: `input` is the `HEADER.PAYLOAD` constructed in sign.c::flux_sign_wrap().   It's a fatal error earlier on if HEADER (which includes mandatory values) cannot be encoded, so by the time mech->sign is called, input must be non-NULL.

I should add a note to sign_mech.h and maybe an assertion in the code to make that invariant more clear.

File: src/lib/sign_munge.c [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    time_t encode_time;
+    int saved_errno;
+
+    assert (sm != NULL);
+
+    /* HEADER.PAYLOAD.SIGNATURE
+     * ^input        ^q
+     */
+    q = strrchr (input, '.');
+    assert (q != NULL);
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/lib/sign_munge.c]: No, sign.c::flux_sign_unwrap() decoded the HEADER and PAYLOAD and bails out on error prior to calling mech->verify, so input will always have at least one of the required periods.

Now that you metion it though, it seems like mech->verify should not be expected to parse HEADER.PAYLOAD.SIGNATURE.  I should change the args so that instead of 
```c
/* verify (required)
 * Given HEADER.PAYLOAD.SIGNATURE and parsed 'header', validate that
 * SIGNATURE is valid over HEADER.PAYLOAD.
 * Return 0 on success, or -1 on error with errno and context error set.
 */
typedef int (*sign_mech_verify_f)(flux_security_t *ctx, const char *input,
                                  const struct kv *header, int flags);
```
we have something like:
```c
/* Validate that 'signature' (a NULL terminated string) is valid over
 * 'input' of size 'inputsz'.
 */
typedef int (*sign_mech_verify_f)(flux_security_t *ctx,
                                  const char *input, int inputsz,
                                  const char *signature,
                                  const struct kv *header, int flags);
```

[the remark this answers is not in the document]
Comment (grondo) [reply src/lib/sign_munge.c]: Similar check is in `header_decode`:
```C
    if (!(p = strchr (input, '.'))) {
        errno = EINVAL;
        return NULL;
    }
    src = input;
    srclen = p - input;

```

Does this need to handle corner cases (period at beginning or end of `input`)?

[the remark this answers is not in the document]
Comment (garlick) [reply src/lib/sign_munge.c]: I think these are OK and are covered by lib/test/sign.c, e.g.

Period at the end "HEADER.PAYLOAD."
> flux_unwrap_fails on missing SIGNATURE with EINVAL

Period at the beginning ".PAYLOAD.SIGNATURE"
> flux_unwrap_fails on empty HEADER with EINVAL
